File: android/src/main/java/com/nitroplay/player/PlayerAppStatePolicy.kt
package com.nitroplay.video.core

internal data class PlayerAppStateSnapshot(
  val playInBackground: Boolean,
  val playWhenInactive: Boolean,
  val isPlaying: Boolean,
  val wasAutoPaused: Boolean,
  val isExternalPlaybackActive: Boolean
)

internal object PlayerAppStatePolicy {
  fun shouldAutoPauseWhenInactive(snapshot: PlayerAppStateSnapshot): Boolean {
    return snapshot.isPlaying &&
      !snapshot.playInBackground &&
      !snapshot.playWhenInactive &&
      !snapshot.isExternalPlaybackActive
  }

  fun shouldAutoPauseWhenEnteringBackground(snapshot: PlayerAppStateSnapshot): Boolean {
    return shouldAutoPauseWhenInactive(snapshot)
  }

  fun shouldResumeWhenEnteringForeground(snapshot: PlayerAppStateSnapshot): Boolean {
    return snapshot.wasAutoPaused
  }

  fun shouldClearAutoPausedAfterBecomeActive(snapshot: PlayerAppStateSnapshot): Boolean {
    return !snapshot.wasAutoPaused || snapshot.isPlaying
  }
}

File: ios/player/PlayerAppStatePolicy.swift
import Foundation

struct PlayerAppStateSnapshot {
  let playInBackground: Bool
  let playWhenInactive: Bool
  let isPlaying: Bool
  let wasAutoPaused: Bool
  let isExternalPlaybackActive: Bool
}

enum PlayerAppStatePolicy {
  static func shouldAutoPauseWhenInactive(_ snapshot: PlayerAppStateSnapshot) -> Bool {
    snapshot.isPlaying
      && !snapshot.playInBackground
      && !snapshot.playWhenInactive
      && !snapshot.isExternalPlaybackActive
  }

  static func shouldAutoPauseWhenEnteringBackground(_ snapshot: PlayerAppStateSnapshot) -> Bool {
    shouldAutoPauseWhenInactive(snapshot)
  }

  static func shouldResumeWhenEnteringForeground(_ snapshot: PlayerAppStateSnapshot) -> Bool {
    snapshot.wasAutoPaused
  }

  static func shouldClearAutoPausedAfterBecomeActive(_ snapshot: PlayerAppStateSnapshot) -> Bool {
    !snapshot.wasAutoPaused || snapshot.isPlaying
  }
}
